Lib/test/test_mmap.py

```diff
@@ -1058,6 +1060,91 @@ def __exit__(self, exc_type, exc_value, traceback):
+    @unittest.skipUnless(hasattr(mmap.mmap, '_protect'), 'test needs debug build')
+    def test_access_violations(self):
+
+        code = textwrap.dedent("""
+            from test.support.os_helper import TESTFN
+            import faulthandler
+            import mmap
+            import os
+            from contextlib import suppress
+
+            # Prevent logging access violations to stderr.
+            faulthandler.disable()
+
+            if os.path.exists(TESTFN):
+                os.unlink(TESTFN)
+
+            PAGESIZE = mmap.PAGESIZE
+            PAGE_NOACCESS = 0x01
+
+            with open(TESTFN, 'bw+') as f:
+                f.write(b'A'* PAGESIZE)
+                f.flush()
+
+                m = mmap.mmap(f.fileno(), PAGESIZE)
+                m._protect(PAGE_NOACCESS, 0, PAGESIZE)
+                with suppress(OSError):
+                    m.read(PAGESIZE)
+                with suppress(OSError):
+                    m.read_byte()
+                with suppress(OSError):
+                    m.readline()
+                with suppress(OSError):
+                    m.write(b'A'* PAGESIZE)
+                with suppress(OSError):
+                    m.write_byte(0)
+                with suppress(OSError):
+                    m[0]  # test mmap_subscript
+                with suppress(OSError):
+                    m[0:10]  # test mmap_subscript
+                with suppress(OSError):
+                    m[0:10:2]  # test mmap_subscript
+                with suppress(OSError):
+                    m[0] = 1
+                with suppress(OSError):
+                    m[0:10] = b'A'* 10
+                with suppress(OSError):
+                    m[0:10:2] = b'A'* 5
+                with suppress(OSError):
+                    m.move(0, 10, 1)
+                with suppress(OSError):
+                    list(m)  # test mmap_item
+        """)
+        rt, stdout, stderr = assert_python_ok("-c", code)
+        self.assertEqual(stdout.strip(), b'')
+        self.assertEqual(stderr.strip(), b'')
+
+    @unittest.skipUnless(os.name == 'nt', 'requires Windows')
+    @unittest.skipUnless(hasattr(mmap.mmap, '_protect'), 'test needs debug build')
+    @unittest.expectedFailure
+    def test_access_violations_fail(self):
+
+        code = textwrap.dedent("""
+            from test.support.os_helper import TESTFN
+            import mmap
+            import os
+            from contextlib import suppress
+
+            if os.path.exists(TESTFN):
+                os.unlink(TESTFN)
+
+            PAGESIZE = mmap.PAGESIZE
+            PAGE_NOACCESS = 0x01
+
+            with open(TESTFN, 'bw+') as f:
+                f.write(b'A'* PAGESIZE)
+                f.flush()
+
+                m = mmap.mmap(f.fileno(), PAGESIZE)
+                m._protect(PAGE_NOACCESS, 0, PAGESIZE)
+                with suppress(OSError):
+                    m.find(b'A')
+        """)
+        assert_python_ok("-c", code)
```

Comment:
I don't see why a test is needed for the `find()` method if it hasn't been made safe. Are you planning to support `[r]find()`? Anyway, this test could use `assert_python_failure()`. Then it wouldn't need `unittest.expectedFailure`.

Reply:
I added the test basically as an anti-test to have higher confidence that the other test is working as it's supposed to.

Reply:
and for `expectedFailure`. I wanted to show that the behaviour is still incorrect. Is there a way to mark a unittest as failing because of outstanding issues?

Reply:
> I would like to add support for (r)find in the future. But it requires more research because of the deep call graph of these functions.

The exception includes the faulting address. Doesn't it suffice to check that the in-page error or access violation occurred in the mapped range of the mmap instance?

Reply:
No, because we don't know what state has been set up that is being bypassed by the exception handler, so we can't safely continue (but maybe we will know once the call graph has been untangled - I haven't looked at it myself).

That said, we almost certainly need the handler to be written in the mmap module, whether it can be untangled or not. So given a lack of choices, it may not be too hard to decide.

Reply:
> No, because we don't know what state has been set up that is being bypassed by the exception handler, so we can't safely continue

Okay, you mean something like if stringlib allocates temporary buffers on the heap that could lead to a memory leak. That would have to be handled locally with a try-finally termination handler. Once an exception handler is found, the subsequent termination handlers get called before the exception handler is called.

Reply:
Right, and any locks that have been taken or released need to be restored, refcounts, etc. We have way too much manually managed state to rely on normal unwinding, unfortunately. If we can target the specific operation (like with memcpy) then it's okay.

Reply:
I don't see an issue like that with `_PyBytes_Find()`, `stringlib_find()`, and `fastsearch()`. It's implemented with automatic variables, no locks, or object references. I think it's okay to implement a safe version of `mmap_gfind()` on Windows. For example:

```c
#if defined(MS_WINDOWS) && !defined(DONT_USE_SEH)
static DWORD
filter_page_exception(EXCEPTION_POINTERS *ptrs, EXCEPTION_RECORD *record)
{
    *record = *ptrs->ExceptionRecord;
    if (record->ExceptionCode == EXCEPTION_IN_PAGE_ERROR ||
        record->ExceptionCode == EXCEPTION_ACCESS_VIOLATION)
    {
        return EXCEPTION_EXECUTE_HANDLER;
    }
    return EXCEPTION_CONTINUE_SEARCH;
}


static DWORD
filter_page_exception_method(mmap_object *self,
                             EXCEPTION_POINTERS *ptrs,
                             EXCEPTION_RECORD *record)
{
    *record = *ptrs->ExceptionRecord;
    if (record->ExceptionCode == EXCEPTION_IN_PAGE_ERROR ||
        record->ExceptionCode == EXCEPTION_ACCESS_VIOLATION)
    {
        uintptr_t address = (uintptr_t)record->ExceptionInformation[1];
        if (address >= (uintptr_t)self->data &&
            address < (uintptr_t)self->data + (uintptr_t)self->size)
        {
            return EXCEPTION_EXECUTE_HANDLER;
        }
    }
    return EXCEPTION_CONTINUE_SEARCH;
}
#endif // defined(MS_WINDOWS) && !defined(DONT_USE_SEH)
```

```c
#if defined(MS_WINDOWS) && !defined(DONT_USE_SEH)
static PyObject *
mmap_gfind(mmap_object *self,
           PyObject *args,
           int reverse)
{
    Py_ssize_t start = self->pos;
    Py_ssize_t end = self->size;
    Py_buffer view;
    Py_ssize_t index;

    CHECK_VALID(NULL);
    if (!PyArg_ParseTuple(args, reverse ? "y*|nn:rfind" : "y*|nn:find",
                          &view, &start, &end))
    {
        return NULL;
    }

    if (start < 0) {
        start += self->size;
    }
    if (start < 0) {
        start = 0;
    }
    else if (start > self->size) {
        start = self->size;
    }

    if (end < 0) {
        end += self->size;
    }
    if (end < 0) {
        end = 0;
    }
    else if (end > self->size) {
        end = self->size;
    }

    CHECK_VALID_OR_RELEASE(NULL, view);

    DWORD error = 0;
    EXCEPTION_RECORD record;

    __try {
        if (end < start) {
            index = -1;
        }
        else if (reverse) {
            assert(0 <= start && start <= end && end <= self->size);
            index = _PyBytes_ReverseFind(self->data + start, end - start,
                                         view.buf, view.len, start);
        }
        else {
            assert(0 <= start && start <= end && end <= self->size);
            index = _PyBytes_Find(self->data + start, end - start,
                                  view.buf, view.len, start);
        }
    }
    __except (filter_page_exception_method(
                self, GetExceptionInformation(), &record))
    {
        assert(record.ExceptionCode == EXCEPTION_IN_PAGE_ERROR ||
               record.ExceptionCode == EXCEPTION_ACCESS_VIOLATION);
        if (record.ExceptionCode == EXCEPTION_IN_PAGE_ERROR) {
            error = LsaNtStatusToWinError(
                        (NTSTATUS)record.ExceptionInformation[2]);
        }
        else {
            error = ERROR_NOACCESS;
        }
    }

    PyBuffer_Release(&view);
    if (error) {
        return PyErr_SetFromWindowsErr(error);
    }
    return PyLong_FromSsize_t(index);
}

#else // !defined(MS_WINDOWS) || defined(DONT_USE_SEH)
static PyObject *
mmap_gfind(mmap_object *self,
           PyObject *args,
           int reverse)
{
    Py_ssize_t start = self->pos;
    Py_ssize_t end = self->size;
    Py_buffer view;
    Py_ssize_t index;

    CHECK_VALID(NULL);
    if (!PyArg_ParseTuple(args, reverse ? "y*|nn:rfind" : "y*|nn:find",
                          &view, &start, &end))
    {
        return NULL;
    }

    if (start < 0) {
        start += self->size;
    }
    if (start < 0) {
        start = 0;
    }
    else if (start > self->size) {
        start = self->size;
    }

    if (end < 0) {
        end += self->size;
    }
    if (end < 0) {
        end = 0;
    }
    else if (end > self->size) {
        end = self->size;
    }

    CHECK_VALID_OR_RELEASE(NULL, view);

    if (end < start) {
        index = -1;
    }
    else if (reverse) {
        assert(0 <= start && start <= end && end <= self->size);
        index = _PyBytes_ReverseFind(self->data + start, end - start,
                                     view.buf, view.len, start);
    }
    else {
        assert(0 <= start && start <= end && end <= self->size);
        index = _PyBytes_Find(self->data + start, end - start,
                              view.buf, view.len, start);
    }

    PyBuffer_Release(&view);
    return PyLong_FromSsize_t(index);
}
#endif // defined(MS_WINDOWS) && !defined(DONT_USE_SEH)
```